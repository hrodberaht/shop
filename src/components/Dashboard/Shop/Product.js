import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { addProductToCart, updateProductInCart } from '../../../store/cart/actionCreators';
import AddedToCart from '../Cart/AddedToCart';
import { getAuthToken, getAuthUserId } from '../../../store/authenticate/selectors';
import { getProductsInCart } from '../../../store/cart/selectors';
import { addToWishlist } from '../../../store/wishlist/actionCerators';

export class Product extends Component {
  state = {
    pcsOrder: 1,
    error: null,
    toggleAddedToCart: false,
  };

  changeInStockToText = (inStock) => {
    if (inStock > 100) return 'full supply';
    if (inStock > 10 && inStock <= 100) return 'medium supply';
    if (inStock > 0 && inStock <= 10) return 'last pieces';

    return 'not available';
  };

  handleChange = (e) => {
    const {
      product: { inStock },
    } = this.props;
    this.setState({ pcsOrder: e.target.value });
    this.checkAmountOrder(+inStock, +e.target.value);
  };

  checkAmountOrder = (inStock, pcsOrder) => {
    if (pcsOrder > inStock) return this.setState({ error: 'Not enough products' });
    if (pcsOrder < 1) return this.setState({ pcsOrder: 1 });
    return this.setState({ error: null });
  };

  calculateTotalPrice = (pcsOrder, price) => pcsOrder * price;

  showAddedToCart = () => {
    const { toggleAddedToCart } = this.state;
    this.setState({ toggleAddedToCart: !toggleAddedToCart });
  };

  moreThanInStock = () => (!!this.state.error);

  redirectToCart = () => this.props.history.push('/cart');

  handleClick = (product) => {
    const {
      addProduct, cart, updateProduct, token,
    } = this.props;
    const inCart = cart.find(prod => prod.productId === product.productId);

    if (inCart) {
      const combProd = Object.assign(inCart, {
        pcsOrder: inCart.pcsOrder + product.pcsOrder,
        totalPrice: inCart.totalPrice + product.totalPrice,
      });
      updateProduct(combProd, token);
    } else {
      addProduct(product, this.props.token);
    }

    this.showAddedToCart();
  };

  handleClickWishlist = (product) => {
    const { addToWish, userId, token } = this.props;
    addToWish(product, userId, token);
  };

  render() {
    const {
      product: {
        id, name, type, price, inStock,
      },
    } = this.props;
    const { pcsOrder, error, toggleAddedToCart } = this.state;
    const totalPrice = this.calculateTotalPrice(pcsOrder, price);
    const productToCart = {
      productId: id,
      name,
      price: +price,
      pcsOrder: +pcsOrder,
      totalPrice,
    };
    return (
      <div>
        <div className="product-item">
          <h3>{name}</h3>
          <h5>{type}</h5>
          <p className="product-item__price">
            <span>$</span>
            {price}
          </p>
          <p>
            <span>In stock:</span>
            <span id="stock-message">{this.changeInStockToText(+inStock)}</span>
          </p>
          <p className="product-item__order">
            Order:
            <input type="number" value={pcsOrder} onChange={this.handleChange} />
            pcs
          </p>
          <p id="stock-error">{error}</p>
          <p className="product-item__total">
            <span>Total price:</span>
            <span id="total-price">
              <span>$</span>
              {totalPrice}
            </span>
          </p>
          <button
            className="btn btn-primary"
            type="button"
            onClick={() => this.handleClick(productToCart)}
            disabled={this.moreThanInStock()}
          >
            Add to cart
          </button>
          <button
            className="btn btn-danger"
            type="button"
            onClick={() => this.handleClickWishlist(productToCart)}
          >
            Wish
          </button>
        </div>
        {toggleAddedToCart && (
          <AddedToCart
            product={productToCart}
            redirect={this.redirectToCart}
            show={this.showAddedToCart}
          />
        )}
      </div>
    );
  }
}

Product.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    type: PropTypes.string,
    price: PropTypes.string,
    inStock: PropTypes.string,
  }).isRequired,
  addProduct: PropTypes.func,
  token: PropTypes.string,
  cart: PropTypes.arrayOf(PropTypes.objectOf),
  updateProduct: PropTypes.func.isRequired,
  userId: PropTypes.string.isRequired,
  addToWish: PropTypes.func.isRequired,
};

Product.defaultProps = {
  addProduct: null,
  token: null,
  cart: [],
};

const mapStateToProps = state => ({
  token: getAuthToken(state),
  cart: getProductsInCart(state),
  userId: getAuthUserId(state),
});
export default withRouter(
  connect(
    mapStateToProps,
    {
      addProduct: addProductToCart,
      updateProduct: updateProductInCart,
      addToWish: addToWishlist,
    },
  )(Product),
);
