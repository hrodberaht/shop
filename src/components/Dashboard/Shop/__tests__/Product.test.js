import { Product } from '../Product';

const defaultProps = {
  product: {
    name: 'test',
    type: 'printer',
    price: 12.0,
    inStock: 30,
  },
  addProduct: jest.fn(),
  token: '1234',
  updateProduct: jest.fn(),
  userId: '12',
  addToWish: jest.fn(),
};
const setup = buildSetup(Product, defaultProps);

describe('<Product />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should show diffrent icon for diffrent value inStock', () => {
    const { wrapper } = setup();
    wrapper.setProps({ product: { inStock: 0 } });
    expect(wrapper.find('#stock-message').props().children).toMatchSnapshot('empty battery');
    wrapper.setProps({ product: { inStock: 10 } });
    expect(wrapper.find('#stock-message').props().children).toMatchSnapshot('quater battery');
    wrapper.setProps({ product: { inStock: 11 } });
    expect(wrapper.find('#stock-message').props().children).toMatchSnapshot('medium battery');
    wrapper.setProps({ product: { inStock: 101 } });
    expect(wrapper.find('#stock-message').props().children).toMatchSnapshot('full battery');
  });

  it('should caltulate total price', () => {
    const { wrapper } = setup();
    wrapper.find('input').simulate('change', { target: { value: 1 } });
    expect(wrapper.find('#total-price').text()).toBe('$12.00');
  });

  it('should check is enough products in stock', () => {
    const { wrapper } = setup();
    wrapper.find('input').simulate('change', { target: { value: 31 } });
    expect(wrapper.find('#stock-error').text()).toBe('Not enough products');
    wrapper.find('input').simulate('change', { target: { value: 29 } });
    expect(wrapper.find('#stock-error').text()).toBe('');
  });
});
