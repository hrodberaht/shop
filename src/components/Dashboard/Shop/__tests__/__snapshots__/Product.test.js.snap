// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> should render without crash 1`] = `
<div>
  <div
    className="product-item"
  >
    <h3>
      test
    </h3>
    <h5>
      printer
    </h5>
    <p
      className="product-item__price"
    >
      <span>
        $
      </span>
      12
    </p>
    <p>
      <span>
        In stock:
      </span>
      <span
        id="stock-message"
      >
        medium supply
      </span>
    </p>
    <p
      className="product-item__order"
    >
      Order:
      <input
        onChange={[Function]}
        type="number"
        value={1}
      />
      pcs
    </p>
    <p
      id="stock-error"
    />
    <p
      className="product-item__total"
    >
      <span>
        Total price:
      </span>
      <span
        id="total-price"
      >
        <span>
          $
        </span>
        12
      </span>
    </p>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add to cart
    </button>
    <button
      className="btn btn-danger"
      onClick={[Function]}
      type="button"
    >
      Wish
    </button>
  </div>
</div>
`;
